### DISCUSSION TITLE
How to set customVehicleCommAdapter‘s orientationangle

### DISCUSSION THREAD

**Discussion by naiveZhou** [openTCS/opentcs#22] [General]
Hi,I send a post to set MyVehicleCommAdapter‘s orientationangle from java web,but it works when i select loopback adapter.
It does not work when I select MyTestAdapter.I appreciate your help.
![image](https://user-images.githubusercontent.com/18117945/174257483-d8659d3c-196e-4495-9de6-f094b79cee6b.png)

![image](https://user-images.githubusercontent.com/18117945/174257164-f237ac4f-fb38-4d7a-8b80-7c39a366517c.png)
![image](https://user-images.githubusercontent.com/18117945/174265739-a258abba-8330-4f19-986e-72de596e13e8.png)

![image](https://user-images.githubusercontent.com/18117945/174265885-0592d7f6-a16a-406f-a1c4-001871c76316.png)

I  set the key and it does not work.
![image](https://user-images.githubusercontent.com/18117945/174265549-495fcce0-f719-4a65-bf82-e86e7b792471.png)

**Comment by martingr** [comment]
Hi @naiveZhou, you refer to setting the orientation angle via/from "Java web". Can you elaborate on what exactly you mean by that?

**Comment by naiveZhou** [comment]
Hi,thank you for your reply. 
I develop a java web with spring boot,and i imported openTCS by maven.
Now,I want to use SetOrientationAngleCommand to set the vehicle's angle,as you konw,it works for virtualvehicle.
Is there any other way that  i can set the real vehicle's angle? I have resolved the problem by socket programing.
getProcessModel().setVehicleOrientationAngle(agvInfo.getTest());//this is my test code.and the agvInfo's data come from socket.
For Example,I use code like this ,but it only works for virtualVehicle.I want to know  if I can make it by RMI interface.
vehicleService.sendCommAdapterCommand(vehicleRef,new SetOrientationAngleCommand(orientationAngle));
I want to know how to make it for  customize vehicle adapter .



  



***@***.***
 
发件人： Martin Grzenia
发送时间： 2022-06-20 14:11
收件人： openTCS/opentcs
抄送： smallzhou; Mention
主题： Re: [openTCS/opentcs] How to set customVehicleCommAdapter‘s orientationangle (Discussion #22)
Hi @naiveZhou, you refer to setting the orientation angle via/from "Java web". Can you elaborate on what exactly you mean by that?

**Reply by martingr** [reply]
I'm not sure I fully understand the setup you are using. You want to use the RMI interface of openTCS to set the orientation angle of a vehicle (via _your own_ vehicle adapter) from the Spring Boot application. Is that correct? I assume the openTCS Kernel application is running separately and not inside the Spring Boot application? What exactly is not working, are there maybe any exceptions thrown? Could you provide a minimal example project that demonstrates the problem? 

Although `SetOrientationAngleCommand` is a loopback communication adapter-specific class, it should also work with other `VehicleCommAdapter` implementations (provided the command is on the openTCS Kernel application's classpath).

**Comment by naiveZhou** [comment]
Yes,You understand very well.and the openTCS Kernel application is running separately and not inside the Spring Boot application.
 for example,i get my vehicle Reference and use sendCommAdapterCommand and i saw nothing exceptions.

this is my Spring Boot application code,
I also created the SetOrientationAngleCommand  class and   import it.But it work when i select virtual vehicle adapter with virtual SetOrientationAngleCommand   and does not work when i use real SetOrientationAngleCommand
import org.opentcs.virtualvehicle.commands.SetOrientationAngleCommand;import org.opentcs.realvehicle.commands.SetOrientationAngleCommand;
 @PostMapping("/setOrientationAngle")
 public void ***@***.*** Double orientationAngle) { 
 KernelServicePortal kernelServicePortal = new KernelServicePortalBuilder("Alice","xyz").build();
 kernelServicePortal.login("127.0.0.1",1099);
 VehicleService vehicleService = kernelServicePortal.getVehicleService(); 
 String vehicleName = "Vehicle-01"; 
 Vehicle vehicle = vehicleService.fetchObject(Vehicle.class,vehicleName); 
 TCSObjectReference vehicleRef = vehicle.getReference(); 
 vehicleService.sendCommAdapterCommand(vehicleRef,new SetOrientationAngleCommand(orientationAngle));
 }

I I don't quite understand that  it should also work with other VehicleCommAdapter implementations (provided the command is on the openTCS Kernel application's classpath).
Do I have to override method for sendCommAdapterCommand  in the OpenTCS ? I override the vehicle adapter and override execute method in the OpenTCS.
@Override
public void execute(AdapterCommand command) {
  getProcessModel().setVehicleOrientationAngle(180.0);
}

when i debug and i saw the the value like this,but the vehicle's angle does not change.





***@***.***
 
发件人： Martin Grzenia
发送时间： 2022-06-22 14:50
收件人： openTCS/opentcs
抄送： smallzhou; Mention
主题： Re: [openTCS/opentcs] How to set customVehicleCommAdapter‘s orientationangle (Discussion #22)
I'm not sure I fully understand the setup you are using. You want to use the RMI interface of openTCS to set the orientation angle of a vehicle (via your own vehicle adapter) from the Spring Boot application. Is that correct? I assume the openTCS Kernel application is running separately and not inside the Spring Boot application? What exactly is not working, are there maybe any exceptions thrown? Could you provide a minimal example project that demonstrates the problem?
Although SetOrientationAngleCommand is a loopback communication adapter-specific class, it should also work with other VehicleCommAdapter implementations (provided the command is on the openTCS Kernel application's classpath).

**Reply by martingr** [reply]
Are there any exceptions thrown in the Kernel application?

Since you created your own command with `org.opentcs.realvehicle.commands.SetOrientationAngleCommand`, you have to ensure that this class is on both the classpath of your Spring Boot application and the classpath of the Kernel application.

Alternatively, you should be able to test it with your vehicle adapter and `org.opentcs.virtualvehicle.commands.SetOrientationAngleCommand`. I would expect this to work.

> Do I have to override method for sendCommAdapterCommand  in the OpenTCS ?

No, you should not have to override this method.

> when i debug and i saw the the value like this,but the vehicle's angle does not change.

So using your `org.opentcs.realvehicle.commands.SetOrientationAngleCommand` you can see `execute(AdapterCommand command)` getting invoked?

**Comment by naiveZhou** [comment]
Thank you very much. Your answer helped me a lot and it work. 
I have to ensure that this class is on both the classpath of your Spring Boot application and the classpath of the Kernel application.
I build the openTCS-Example-CommAdapter jar and maven install in the local reposity ,then i import the dependency  of openTCS-Example-CommAdapter's jar in the Spring Boot application.
 


***@***.***
 
发件人： Martin Grzenia
发送时间： 2022-06-23 14:39
收件人： openTCS/opentcs
抄送： smallzhou; Mention
主题： Re: [openTCS/opentcs] How to set customVehicleCommAdapter‘s orientationangle (Discussion #22)
Are there any exceptions thrown in the Kernel application?
Since you created your own command with org.opentcs.realvehicle.commands.SetOrientationAngleCommand, you have to ensure that this class is on both the classpath of your Spring Boot application and the classpath of the Kernel application.
Alternatively, you should be able to test it with your vehicle adapter and org.opentcs.virtualvehicle.commands.SetOrientationAngleCommand. I would expect this to work.
Do I have to override method for sendCommAdapterCommand in the OpenTCS ?
No, you should not have to override this method.
when i debug and i saw the the value like this,but the vehicle's angle does not change.
So using your org.opentcs.virtualvehicle.commands.SetOrientationAngleCommand you can see this method getting invoked?
